File: benchmarking/learning/features.py
"""Deterministic offline mirror of Bouncer's Go feature extractor."""

from __future__ import annotations

import json
import math
from collections.abc import Mapping
from typing import Any

FEATURE_SCHEMA_VERSION = "0.1.0"
OPERATION_CLASSES = (
    "filesystem.read",
    "filesystem.write",
    "filesystem.delete",
    "state.validate",
    "state.backup",
    "command.run",
    "service.deploy",
    "task.complete",
)
_BASE_FEATURES = (
    "turn_fraction",
    "remaining_turn_fraction",
    "mutation_fraction",
    "remaining_mutation_fraction",
    "completed_operation_count",
    "file_count",
    "constraint_feedback_count",
    "recent_rejection_count",
    "no_progress_streak",
    "candidate_mutating",
    "dependency_count",
    "dependency_satisfaction_ratio",
    "argument_count",
    "argument_bytes_log1p",
    "target_depth",
    "calibrated_latency_log1p",
    "calibrated_cost_log1p",
    "calibrated_safety_risk",
    "mutation_budget_interaction",
    "risk_mutation_interaction",
    "transition_log_probability",
    "transition_unseen",
)
FEATURE_NAMES = _BASE_FEATURES + tuple(
    name
    for operation in OPERATION_CLASSES
    for name in (f"operation={operation}", f"previous={operation}")
)


def extract_features(
    context: Mapping[str, Any],
    candidate: Mapping[str, Any],
    transition_prior: Mapping[str, Any],
) -> dict[str, float]:
    """Return the exact feature vector expected by the Go runtime."""
    features = dict.fromkeys(FEATURE_NAMES, 0.0)
    max_turns = max(_integer(context, "max_turns"), 1)
    turn = min(max(_integer(context, "turn"), 0), max_turns)
    state = _mapping(context, "state")
    policy = _mapping(context, "policy")
    max_mutations = max(_integer(policy, "max_mutations"), 1)
    mutations = min(max(_integer(state, "mutation_count"), 0), max_mutations)
    completed = _string_list(state, "completed_operations")
    files = _mapping(state, "files")
    feedback = state.get("constraint_feedback", [])
    if not isinstance(feedback, list):
        raise ValueError("state.constraint_feedback must be an array")

    features["turn_fraction"] = turn / max_turns
    features["remaining_turn_fraction"] = max(max_turns - turn, 0) / max_turns
    features["mutation_fraction"] = mutations / max_mutations
    features["remaining_mutation_fraction"] = max(max_mutations - mutations, 0) / max_mutations
    features["completed_operation_count"] = _bounded_count(len(completed))
    features["file_count"] = _bounded_count(len(files))
    features["constraint_feedback_count"] = _bounded_count(len(feedback))
    features["recent_rejection_count"] = _bounded_count(_integer(context, "recent_rejections"))
    features["no_progress_streak"] = _bounded_count(_integer(context, "no_progress_streak"))

    operation = _string(candidate, "operation_class")
    arguments = _mapping(candidate, "arguments")
    dependencies = _string_list(candidate, "declared_dependencies")
    target = _string(candidate, "target")
    objectives = _mapping(candidate, "routing_objectives")
    mutating = operation in {"filesystem.write", "filesystem.delete", "service.deploy"}
    features["candidate_mutating"] = float(mutating)
    features["dependency_count"] = _bounded_count(len(dependencies))
    features["dependency_satisfaction_ratio"] = (
        sum(dependency in set(completed) for dependency in dependencies) / len(dependencies)
        if dependencies
        else 1.0
    )
    features["argument_count"] = _bounded_count(len(arguments))
    encoded_arguments = json.dumps(
        arguments,
        sort_keys=True,
        separators=(",", ":"),
        ensure_ascii=False,
    ).encode()
    features["argument_bytes_log1p"] = math.log1p(min(len(encoded_arguments), 1 << 20))
    features["target_depth"] = _bounded_count(target.count("/") + 1)
    features["calibrated_latency_log1p"] = math.log1p(
        _non_negative_number(objectives, "latency_ms")
    )
    features["calibrated_cost_log1p"] = math.log1p(_non_negative_number(objectives, "cost_units"))
    features["calibrated_safety_risk"] = _probability(objectives, "safety_risk")
    features["mutation_budget_interaction"] = (
        features["candidate_mutating"] * features["remaining_mutation_fraction"]
    )
    features["risk_mutation_interaction"] = (
        features["candidate_mutating"] * features["calibrated_safety_risk"]
    )

    previous = context.get("previous_operation") or "START"
    if not isinstance(previous, str):
        raise ValueError("previous_operation must be a string")
    probability, seen = transition_probability(transition_prior, previous, operation)
    features["transition_log_probability"] = max(math.log(probability), -20.0)
    features["transition_unseen"] = float(not seen)
    if operation in OPERATION_CLASSES:
        features[f"operation={operation}"] = 1.0
    if previous in OPERATION_CLASSES:
        features[f"previous={previous}"] = 1.0
    return features


def transition_probability(
    prior: Mapping[str, Any], previous: str, operation: str
) -> tuple[float, bool]:
    """Return a smoothed transition probability and whether it was observed."""
    fallback = _probability(prior, "fallback_probability", exclusive_zero=True)
    probabilities = _mapping(prior, "probabilities")
    next_operations = probabilities.get(previous)
    if isinstance(next_operations, Mapping) and operation in next_operations:
        value = next_operations[operation]
        if isinstance(value, bool) or not isinstance(value, int | float):
            raise ValueError("transition probability must be numeric")
        probability = float(value)
        if not 0 < probability <= 1 or not math.isfinite(probability):
            raise ValueError("transition probability must be in (0,1]")
        return probability, True
    return fallback, False


def _mapping(value: Mapping[str, Any], key: str) -> Mapping[str, Any]:
    result = value.get(key)
    if not isinstance(result, Mapping):
        raise ValueError(f"{key} must be an object")
    return result


def _string(value: Mapping[str, Any], key: str) -> str:
    result = value.get(key)
    if not isinstance(result, str) or not result:
        raise ValueError(f"{key} must be a non-empty string")
    return result


def _string_list(value: Mapping[str, Any], key: str) -> list[str]:
    result = value.get(key)
    if not isinstance(result, list) or any(not isinstance(item, str) for item in result):
        raise ValueError(f"{key} must be an array of strings")
    return result


def _integer(value: Mapping[str, Any], key: str) -> int:
    result = value.get(key, 0)
    if isinstance(result, bool) or not isinstance(result, int):
        raise ValueError(f"{key} must be an integer")
    return result


def _non_negative_number(value: Mapping[str, Any], key: str) -> float:
    result = value.get(key)
    if isinstance(result, bool) or not isinstance(result, int | float):
        raise ValueError(f"{key} must be numeric")
    number = float(result)
    if not math.isfinite(number) or number < 0:
        raise ValueError(f"{key} must be finite and non-negative")
    return number


def _probability(value: Mapping[str, Any], key: str, *, exclusive_zero: bool = False) -> float:
    number = _non_negative_number(value, key)
    if number > 1 or (exclusive_zero and number == 0):
        interval = "(0,1]" if exclusive_zero else "[0,1]"
        raise ValueError(f"{key} must be in {interval}")
    return number


def _bounded_count(value: int) -> float:
    return math.log1p(max(value, 0))

File: internal/learning/features.go
package learning

import (
	"encoding/json"
	"math"
	"strings"

	"bouncer/internal/action"
)

var operationClasses = []string{
	"filesystem.read",
	"filesystem.write",
	"filesystem.delete",
	"state.validate",
	"state.backup",
	"command.run",
	"service.deploy",
	"task.complete",
}

var featureNames = func() []string {
	names := []string{
		"turn_fraction",
		"remaining_turn_fraction",
		"mutation_fraction",
		"remaining_mutation_fraction",
		"completed_operation_count",
		"file_count",
		"constraint_feedback_count",
		"recent_rejection_count",
		"no_progress_streak",
		"candidate_mutating",
		"dependency_count",
		"dependency_satisfaction_ratio",
		"argument_count",
		"argument_bytes_log1p",
		"target_depth",
		"calibrated_latency_log1p",
		"calibrated_cost_log1p",
		"calibrated_safety_risk",
		"mutation_budget_interaction",
		"risk_mutation_interaction",
		"transition_log_probability",
		"transition_unseen",
	}
	for _, operation := range operationClasses {
		names = append(names, "operation="+operation, "previous="+operation)
	}
	return names
}()

// FeatureNames returns the stable ordered feature schema.
func FeatureNames() []string {
	return append([]string(nil), featureNames...)
}

func featureNameSet() map[string]struct{} {
	result := make(map[string]struct{}, len(featureNames))
	for _, name := range featureNames {
		result[name] = struct{}{}
	}
	return result
}

func extractFeatures(
	context Context,
	candidate action.ScoredCandidate,
	prior TransitionPrior,
) map[string]float64 {
	features := make(map[string]float64, len(featureNames))
	for _, name := range featureNames {
		features[name] = 0
	}
	maxTurns := max(context.MaxTurns, 1)
	turn := min(max(context.Turn, 0), maxTurns)
	features["turn_fraction"] = float64(turn) / float64(maxTurns)
	features["remaining_turn_fraction"] = float64(max(maxTurns-turn, 0)) / float64(maxTurns)
	maxMutations := max(context.Policy.MaxMutations, 1)
	mutations := min(max(context.State.MutationCount, 0), maxMutations)
	features["mutation_fraction"] = float64(mutations) / float64(maxMutations)
	features["remaining_mutation_fraction"] = float64(max(maxMutations-mutations, 0)) / float64(maxMutations)
	features["completed_operation_count"] = boundedCount(len(context.State.CompletedOperations))
	features["file_count"] = boundedCount(len(context.State.Files))
	features["constraint_feedback_count"] = boundedCount(len(context.State.ConstraintFeedback))
	features["recent_rejection_count"] = boundedCount(context.RecentRejections)
	features["no_progress_streak"] = boundedCount(context.NoProgressStreak)
	mutating := isMutating(candidate.Candidate.OperationClass)
	features["candidate_mutating"] = boolFloat(mutating)
	features["dependency_count"] = boundedCount(len(candidate.Candidate.DeclaredDependencies))
	features["dependency_satisfaction_ratio"] = dependencyRatio(
		candidate.Candidate.DeclaredDependencies,
		context.State.CompletedOperations,
	)
	features["argument_count"] = boundedCount(len(candidate.Candidate.Arguments))
	if encoded, err := json.Marshal(candidate.Candidate.Arguments); err == nil {
		features["argument_bytes_log1p"] = math.Log1p(float64(min(len(encoded), 1<<20)))
	}
	features["target_depth"] = boundedCount(strings.Count(candidate.Candidate.Target, "/") + 1)
	features["calibrated_latency_log1p"] = math.Log1p(candidate.RoutingObjectives.LatencyMS)
	features["calibrated_cost_log1p"] = math.Log1p(candidate.RoutingObjectives.CostUnits)
	features["calibrated_safety_risk"] = candidate.RoutingObjectives.SafetyRisk
	features["mutation_budget_interaction"] = features["candidate_mutating"] * features["remaining_mutation_fraction"]
	features["risk_mutation_interaction"] = features["candidate_mutating"] * features["calibrated_safety_risk"]
	probability, seen := transitionProbability(prior, context.PreviousOperation, candidate.Candidate.OperationClass)
	features["transition_log_probability"] = math.Max(math.Log(probability), -20)
	features["transition_unseen"] = boolFloat(!seen)
	features["operation="+candidate.Candidate.OperationClass] = 1
	if context.PreviousOperation != "" {
		features["previous="+context.PreviousOperation] = 1
	}
	return features
}

func transitionProbability(prior TransitionPrior, previous, next string) (float64, bool) {
	if previous == "" {
		previous = "START"
	}
	if transitions, ok := prior.Probabilities[previous]; ok {
		if probability, found := transitions[next]; found {
			return probability, true
		}
	}
	return prior.FallbackProbability, false
}

func dependencyRatio(dependencies, completed []string) float64 {
	if len(dependencies) == 0 {
		return 1
	}
	set := make(map[string]struct{}, len(completed))
	for _, operation := range completed {
		set[operation] = struct{}{}
	}
	satisfied := 0
	for _, dependency := range dependencies {
		if _, ok := set[dependency]; ok {
			satisfied++
		}
	}
	return float64(satisfied) / float64(len(dependencies))
}

func isMutating(operation string) bool {
	return operation == "filesystem.write" || operation == "filesystem.delete" || operation == "service.deploy"
}

func boundedCount(value int) float64 {
	return math.Log1p(float64(max(value, 0)))
}

func boolFloat(value bool) float64 {
	if value {
		return 1
	}
	return 0
}
